liquibase-core/src/main/java/liquibase/change/core/ExecuteShellCommandChange.java

```diff
@@ -265,7 +266,8 @@ public void run() {
                 // check if we timed out or not
                 if (timedOut.get()) {
                     String timeoutStr = timeout != null ? timeout : timeoutInMillis + " ms";
-                    throw new TimeoutException("Process timed out (" + timeoutStr + ")");
+                    LogFactory.getInstance().getLog().severe("Process timed out (" + timeoutStr + ")");
+                    return KILLED_PROCESS_RETURN_CODE;
```

Comment:
For me, this return looks not completely 'fair' (we do return sigterm without any guarantee from OS side, that process destroy task, scheduled within timer, actually 'did its work'. From this piece of code, it seems the only way to guarantee it is to wait until 'ret' value of process is indeed initialized.  
Overall, this method seems to be should refactored a little bit:
1. Rename 'waitForOrKill' to 'waitWithTimeOut'
2. Remove condition `timedOut.get()` (and inner logic), as timeout is already checked via scheduled in timer process destroy task
3. Remove this weird and useless `while(!stop)` , leave `try` block as it is (method waitFor() of ProcessImpl has native waitForInterruptibly() which ALREADY contains needed all awaiting logic for external process)
4. Check 'ret' value, if it is 143, then log ("process interrupted due to timeout...")

I suggest to do it like this:

   ```
 private int waitWithTimeout(final Process process, final long timeoutInMillis) throws ExecutionException, TimeoutException {
        int ret = -1;
        final AtomicBoolean timedOut = new AtomicBoolean(false);
        Timer timer = new Timer();
        if (timeoutInMillis > 0) {
            timer.schedule(new TimerTask() {
                @Override
                public void run() {
                    // timed out
                    timedOut.set(true);
                    process.destroy();
                }
            }, timeoutInMillis);
        }

        String timeoutStr = timeout != null ? timeout : timeoutInMillis + " ms";
        try {
            ret = process.waitFor();
            // check if we timed out or not
        } catch (InterruptedException interruptedException) {
            // check again
            // Restore interrupted state...
            LogFactory.getInstance().getLog().severe("Process interrupted due to ", interruptedException);
            Thread.currentThread().interrupt();
        } finally {
            timer.cancel();
        }

        if (KILLED_PROCESS_RETURN_CODE == ret) {
            LogFactory.getInstance().getLog().severe("Process timed out (" + timeoutStr + ")");
        }

        return ret;
    }
```

Reply:
And even if we keep `while (!stop)` as a some `retry` mechanism, at least we should add some Thread.sleep(timeout/10) to avoid redundant bother of OS Process Planner (and useless OS 'context switching' if there is not too much CPU on running machine)

Reply:
Hi @v-petrovych, I agree that we shouldn't return hardcoded 143, as it can also be used on Windows OS and or in some cases return code in case of process kill can be different, so I made the change to return original process exit code in case of time out.

1. Regarding method rename - I think that the name is good, as we kill the process if it times out. Why do you think that we don't kill it?
2. Regarding timedOut.get() removal - removed it and added logging of error message after destroy method to be able to track if it was timed out.
3. While loop with stop is not for awaiting, it is for an additional retry which happens only once when interrupted exception happens, I wasn't able to find info why exactly we need it, but it was definitely done intentionally, so we shouldn't remove it.
4. This one is also not needed as we add error log in time test directly
5. Regarding your latest comment about Thread.sleep - we don't need it since retry happens only once and we don't need to wait until it is done, so no resources are wasted

Reply:
LGTM!